### PR TITLE
refine AliasType deducing when planning

### LINKED ISSUE #4968: Alias validation error
**Please check the FAQ documentation before raising an issue**

**Your Environments (__required__)**
nebula latest version
**Describe the bug (__required__)**
The variable `curr` is a node, but was incorrectly set to the path type.
**How To Reproduce(__required__)**
```
match p=allShortestPaths((v1)-[:like*1..3]-(v2)) where id(v1)=="Tim Duncan" and id(v2)=="Tony Parker" with nodes(p) as pathNodes WITH   [n IN pathNodes | id(n)] AS personIdsInPath,   [idx IN range(1, size(pathNodes)-1) | [prev IN [pathNodes[idx-1]] | [curr IN [pathNodes[idx]] | [prev, curr]]]] AS vertList UNWIND vertList AS c WITH c[0][0][0] AS prev, c[0][0][1] AS curr, personIdsInPath OPTIONAL MATCH (curr)<-[e:like]-(:player)-[:serve]->(:team)-[:teammate]->(prev) RETURN count(e) AS cnt1, prev, curr, personIdsInPath
```
Error:  SemanticError: curr binding to different type: Node vs Path

**Expected behavior**
Verify variable types correctly.

**Additional context**
We'd be better off doing some refactoring of our existing implementation of alias management.

### PR DESCRIPTION
**Pull request by codesigner**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 
fix #4968 

#### Description:

- The AliasType deducing during planning time is not good enough, but we do need some check during planning for join on Node or Edge, like Edge cannot be join with EdgeList, etc.
- But, currently, we cannot deduce accurate deducing every expression, due to the lack of semantic analyze of Expression, and schema-free feature of open cypher, deduced to kRuntime (leave type check to runtime to check the validity)

Changes
- A DeduceAliasTypeVisitor was added, intened be used to deduce some AliasType from expression, will escape to kRuntime if it cannot.
- Relax join key check when meet kRuntime AliasType in MatchPlanner.
- kDefault changed to kRuntime, which is more accurate.

## Checklist:
Tests:
- [x] TCK
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Review by yixinglu — APPROVED** [review]

File: src/graph/visitor/DeduceAliasTypeVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  auto funName = expr->name();
+  if (funName == "nodes") {
+    outputType_ = AliasType::kNodeList;
+  } else if (funName == "relationships") {
+    outputType_ = AliasType::kEdgeList;
+  } else if (funName == "reversepath") {
+    outputType_ = AliasType::kPath;
+  } else if (funName == "startnode" || funName == "endnode") {
+    outputType_ = AliasType::kNode;
+  }
```
Comment (czpmango) [inline src/graph/visitor/DeduceAliasTypeVisitor.cpp:61]: There may be case-sensitive issues with function names.

File: src/graph/visitor/DeduceAliasTypeVisitor.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  void visit(ConstantExpression *) override {}
+  void visit(UnaryExpression *) override {}
+  void visit(TypeCastingExpression *) override {}
+  void visit(LabelExpression *) override {}
+  void visit(LabelAttributeExpression *) override {}
+  void visit(ArithmeticExpression *) override {}
+  void visit(RelationalExpression *) override {}
+  void visit(AttributeExpression *) override {}
+  void visit(LogicalExpression *) override {}
+  void visit(AggregateExpression *) override {}
```
Comment (czpmango) [inline src/graph/visitor/DeduceAliasTypeVisitor.h:54]: Do these expressions need to be processed, such as `match (v) with collect(v)[0] as n match ()-[v]-() return v`.

File: src/graph/visitor/DeduceAliasTypeVisitor.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  void visit(TagPropertyExpression *) override {}
+  void visit(EdgePropertyExpression *) override {}
+  void visit(InputPropertyExpression *) override {}
+  void visit(VariablePropertyExpression *) override {}
+  void visit(DestPropertyExpression *) override {}
+  void visit(SourcePropertyExpression *) override {}
+  void visit(EdgeSrcIdExpression *) override {}
+  void visit(EdgeTypeExpression *) override {}
+  void visit(EdgeRankExpression *) override {}
+  void visit(EdgeDstIdExpression *) override {}
```
Comment (czpmango) [inline src/graph/visitor/DeduceAliasTypeVisitor.h:71]: catalog needs to be introduced to enhance the inference of these expression types.
We can do that later.

**Review by czpmango — APPROVED** [review]
LGTM

File: src/graph/visitor/DeduceAliasTypeVisitor.h (hunk 2 above)
Comment (codesigner) [reply src/graph/visitor/DeduceAliasTypeVisitor.h:51]: > Do these expressions need to be processed, such as `match (v) with collect(v)[0] as n match ()-[v]-() return v`.

Some do has the potential, may retrieve more info  during planning phase.

File: src/graph/visitor/DeduceAliasTypeVisitor.cpp (hunk 1 above)
Comment (codesigner) [reply src/graph/visitor/DeduceAliasTypeVisitor.cpp:54]: fixed

**Review by czpmango — APPROVED** [review]

File: src/graph/visitor/DeduceAliasTypeVisitor.h (hunk 2 above)
Comment (czpmango) [reply src/graph/visitor/DeduceAliasTypeVisitor.h:54]: ACK.

File: src/graph/visitor/DeduceAliasTypeVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+namespace graph {
+
+DeduceAliasTypeVisitor::DeduceAliasTypeVisitor(QueryContext *qctx,
+                                               ValidateContext *vctx,
+                                               GraphSpaceID space,
+                                               AliasType inputType)
+    : qctx_(qctx), vctx_(vctx), space_(space), inputType_(inputType) {
+  UNUSED(qctx_);
+  UNUSED(vctx_);
+  UNUSED(space_);
```
[the remark this answers is not in the document]
Comment (codesigner) [reply src/graph/visitor/DeduceAliasTypeVisitor.cpp:33]: > Delete these members.

This DeduceAliasTypeVisitor may be refined in the future, these member may be used by then.

File: src/graph/planner/match/MatchPlanner.h [hunk 5]
```diff
@@ -23,7 +23,8 @@ class MatchPlanner final : public Planner {
   StatusOr<SubPlan> transform(AstContext* astCtx) override;
 
  private:
-  static constexpr std::array AliasTypeName = {"Node", "Edge", "Path", "EdgeList", "Default"};
+  static constexpr std::array AliasTypeName = {
+      "Node", "Edge", "Path", "NodeList", "EdgeList", "Default"};
```
Comment (yixinglu) [inline src/graph/planner/match/MatchPlanner.h:27]: default -> runtime ?

**Review by czpmango — DISMISSED** [review]

File: src/graph/planner/match/MatchPlanner.h (hunk 5 above)
Comment (codesigner) [reply src/graph/planner/match/MatchPlanner.h:27]: > default -> runtime ?

Good catch

**Comment by czpmango** [comment]
Add test case:
```
match p=(a:player)-[e:like*1..3]->(b) where b.player.age>42 
with relationships(p)[1] AS e1 match (b)-[:serve]->(c) where c.team.name>"S" and (b)-[e1]->() 
return c
```

**Comment by codesigner** [comment]
> Add test case:
> 
> ```
> match p=(a:player)-[e:like*1..3]->(b) where b.player.age>42 
> with relationships(p)[1] AS e1 match (b)-[:serve]->(c) where c.team.name>"S" and (b)-[e1]->() 
> return c
> ```

added

**Comment by yixinglu** [comment]
@codesigner plz fix the compile errors.

**Review by yixinglu — APPROVED** [review]

**Review by czpmango — DISMISSED** [review]
LGTM

**Review by yixinglu — APPROVED** [review]

**Review by czpmango — APPROVED** [review]

**Review by jievince — DISMISSED** [review]

**Review by czpmango — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**